### PR TITLE
Add 2, update 2, ai dev new templates to allow use of OpenAI JS library, use of OpenAI.com keys, and Assistants API

### PR DESCRIPTION
**Pull request by robch**
Updated these `ai dev new` templates (now uses OpenAI JS library, can use OpenAI.com or AOAI connection info)
* openai-chat-streaming-js
* openai-webpage-js

Added these `ai dev new` templates (uses OpenAI JS library, can use OpenAI.com or AOAI connection info)
* openai-chat-w-asst-w-streaming
* openaiwhat-w-asst-webpage

### CODE REVIEW

File: src/ai/.x/templates/openai-chat-w-asst-w-streaming-js/Main.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  await assistant.getThreadMessages((role, content) => {
+    role = role.charAt(0).toUpperCase() + role.slice(1);
+    process.stdout.write(`${role}: ${content}`);
+  });
+
+  // Loop until the user types 'exit'
+  while (true) {
+
+    // Get user input
+    const input = await new Promise(resolve => rl.question('User: ', resolve));
```
Comment (manekinekko) [inline src/ai/.x/templates/openai-chat-w-asst-w-streaming-js/Main.js:53]:
```suggestion
    const input = await rl.question('User: ');
```

File: src/ai/.x/templates/openai-chat-w-asst-w-streaming-js/Main.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
+<#@ template hostspecific="true" #>
```
Comment (manekinekko) [inline src/ai/.x/templates/openai-chat-w-asst-w-streaming-js/Main.js:1]: I'd rename `Main.js` to `main.js`

File: src/ai/.x/templates/openai-chat-w-asst-webpage-js/src/OpenAIAssistantsStreamingClass.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+const { OpenAI } = require('openai');
+
+class <#= ClassName #> {
+
+  // Create the class using the Azure OpenAI API, which requires a different setup, baseURL, api-key headers, and query parameters
+  static createUsingAzure(azureOpenAIAPIVersion, azureOpenAIEndpoint, azureOpenAIKey, azureOpenAIDeploymentName, openAIAssistantId) {
+    console.log("Using Azure OpenAI API...");
+    return new <#= ClassName #>(openAIAssistantId,
+      new OpenAI({
+        apiKey: azureOpenAIKey,
```
Comment (manekinekko) [inline src/ai/.x/templates/openai-chat-w-asst-webpage-js/src/OpenAIAssistantsStreamingClass.js:13]: We should add a disclaimer asking users not to deploy their OpenAI API key on client side, as described [here](https://help.openai.com/en/articles/5112595-best-practices-for-api-key-safety).

File: src/ai/.x/templates/openai-chat-w-asst-webpage-js/src/script.js [hunk 4]
```diff
@@ -0,0 +1,512 @@
...
+  };
+}
+
+const titleUntitled = 'Untitled';
+
+function ThreadItem(id, created, metadata) {
+  this.id = id;
+  this.created = created;
+  this.metadata = metadata;
+}
```
Comment (manekinekko) [inline src/ai/.x/templates/openai-chat-w-asst-webpage-js/src/script.js:239]: This function can be converted into a Class.

File: src/ai/.x/templates/openai-chat-w-asst-webpage-js/webpack.config.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
+const path = require('path');
+const webpack = require('webpack');
+const Dotenv = require('dotenv-webpack');
+
+module.exports = {
+  entry: './src/script.js',
```
Comment (manekinekko) [inline src/ai/.x/templates/openai-chat-w-asst-webpage-js/webpack.config.js:6]:
```suggestion
  mode: 'development',
  entry: './src/script.js',
```

File: src/ai/.x/templates/openai-chat-w-asst-webpage-js/package.json [hunk 6]
```diff
@@ -0,0 +1,22 @@
+{
+  "name": "chat-interface",
+  "version": "1.0.0",
+  "description": "Chat Interface with OpenAI",
+  "main": "script.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
```
Comment (manekinekko) [inline src/ai/.x/templates/openai-chat-w-asst-webpage-js/package.json:7]:
```suggestion
    "test": "echo \"Error: no test specified\" && exit 1",
    "webpack": "webpack"
```

File: src/ai/.x/templates/openai-chat-w-asst-webpage-js/README.md [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+User input is typed into a text box and added to the conversation as a message inside a chat panel. The panel scrolls up and the computer responds with streaming text output into another message in the chat panel. There is a left nav that has a "new chat" button and has a spot for future expansion w/ a list of historical chats.
+
+## Setup
+
+To build the website, run the following commands:
+
+```bash
+npm install
+npx webpack
```
Comment (manekinekko) [inline src/ai/.x/templates/openai-chat-w-asst-webpage-js/README.md:13]:
```suggestion
npm run webpack
```

**Comment by manekinekko** [comment]
These are the most important changes I had related to JavaScript and Node.js convention/practices

File: src/ai/.x/templates/openai-chat-w-asst-w-streaming-js/Main.js [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+<#@ parameter type="System.String" name="OPENAI_API_KEY" #>
+<#@ parameter type="System.String" name="OPENAI_MODEL_NAME" #>
+<#@ parameter type="System.String" name="OPENAI_ORG_ID" #>
+<#@ parameter type="System.String" name="AZURE_OPENAI_API_KEY" #>
+<#@ parameter type="System.String" name="AZURE_OPENAI_API_VERSION" #>
+<#@ parameter type="System.String" name="AZURE_OPENAI_CHAT_DEPLOYMENT" #>
+<#@ parameter type="System.String" name="AZURE_OPENAI_ASSISTANT_ID" #>
+const { <#= ClassName #> } = require("./OpenAIAssistantsStreamingClass");
+
+const readline = require('readline');
```
[the remark this answers is not in the document]
Comment (robch) [reply src/ai/.x/templates/openai-chat-w-asst-w-streaming-js/Main.js:14]: done.

File: src/ai/.x/templates/openai-chat-w-asst-w-streaming-js/Main.js (hunk 1 above)
Comment (robch) [reply src/ai/.x/templates/openai-chat-w-asst-w-streaming-js/Main.js:53]: done.

File: src/ai/.x/templates/openai-chat-w-asst-w-streaming-js/Main.js (hunk 2 above)
Comment (robch) [reply src/ai/.x/templates/openai-chat-w-asst-w-streaming-js/Main.js:1]: done.

File: src/ai/.x/templates/openai-chat-w-asst-webpage-js/webpack.config.js (hunk 5 above)
Comment (robch) [reply src/ai/.x/templates/openai-chat-w-asst-webpage-js/webpack.config.js:6]: done.

File: src/ai/.x/templates/openai-chat-w-asst-webpage-js/package.json (hunk 6 above)
Comment (robch) [reply src/ai/.x/templates/openai-chat-w-asst-webpage-js/package.json:7]: done.

File: src/ai/.x/templates/openai-chat-w-asst-webpage-js/README.md (hunk 7 above)
Comment (robch) [reply src/ai/.x/templates/openai-chat-w-asst-webpage-js/README.md:13]: done.

File: src/ai/.x/templates/openai-chat-w-asst-webpage-js/src/script.js (hunk 4 above)
Comment (robch) [reply src/ai/.x/templates/openai-chat-w-asst-webpage-js/src/script.js:239]: i like this feedback, but i'll save this for a future PR where I do this for more samples in javascript across the board. Thanks!

**Review by rhurey — APPROVED** [review]

File: src/ai/.x/templates/openai-chat-w-asst-webpage-js/src/OpenAIAssistantsStreamingClass.js (hunk 3 above)
Comment (robch) [reply src/ai/.x/templates/openai-chat-w-asst-webpage-js/src/OpenAIAssistantsStreamingClass.js:13]: done.

Comment (robch) [reply src/ai/.x/templates/openai-chat-w-asst-webpage-js/src/OpenAIAssistantsStreamingClass.js:13]: done.

**Pull request merged**